server/src/webserver.rs

```diff
@@ -110,13 +110,51 @@ async fn save_source(
     //     source code.</p>
     language_lexers_compiled: web::Data<LanguageLexersCompiled<'_>>,
 ) -> impl Responder {
+
+    // <p>Takes the source file and the lexer and saves the source as a string.
+    // </p>
+    let (file_contents, _) = save_source_as_string(client_source_file, language_lexers_compiled);
+
+    // <h3>Save file</h3>
+    // <p>Save this string to a file. Add a leading slash for Linux: this
+    //     changes from&nbsp;<code>foo/bar.c</code> to <code>/foo/bar.c</code>.
+    //     Windows already starts with a drive letter, such as
+    //     <code>C:\foo\bar.c</code>, so no changes are needed.</p>
+    let save_file_path = if cfg!(windows) {
+        "".to_string()
+    } else {
+        "/".to_string()
+    } + &encoded_path;
+    match fs::write(save_file_path.to_string(), file_contents).await {
+        Ok(v) => v,
+        Err(err) => {
+            return save_source_response(
+                false,
+                &format!("Unable to save file {}: {}.", save_file_path, err),
+            )
+        }
+    }
+
+    save_source_response(true, "")
+}
+
+    // <p>This function takes in a file with code and doc blocks and outputs a
+    //     string of the contents for testing.</p>
+fn save_source_as_string(
+    // <p>The file to save plus metadata, stored in the
+    //     <code>ClientSourceFile</code></p>
+    client_source_file: web::Json<ClientSourceFile>,
+    // <p>Lexer info, needed to transform the <code>ClientSourceFile</code> into
+    //     source code.</p>
+    language_lexers_compiled: web::Data<LanguageLexersCompiled<'_>>,
+) -> (String, impl Responder) {
     // <p>Given the mode, find the lexer.</p>
     let lexer = match language_lexers_compiled
         .map_mode_to_lexer
         .get(client_source_file.metadata.mode.as_str())
     {
         Some(v) => v,
-        None => return save_source_response(false, "Invalid mode"),
+        None => return (String::new(), save_source_response(false, "Invalid mode")),
```

Comment:
OK for this PR, but we need to think about error handling here. I'm probably recommend returning a [result](https://doc.rust-lang.org/std/result/), then unwrapping that in the calling function.